tiledb/sm/filesystem/test/unit_ls_filtered.cc

```diff
@@ -77,48 +119,583 @@ class VFSTest {
+        parents.push_back(absuri);
+      } while (absuri != vfs_test.temp_dir_);
+
+      parents.pop_back(); /* temp_dir_ */
+      while (!parents.empty()) {
+        REQUIRE(vfs_test.vfs_.create_dir(parents.back()).ok());
+        parents.pop_back();
+      }
+    }
+    REQUIRE(vfs_test.vfs_.touch(tiledb::sm::URI(abspath.string())).ok());
+    std::filesystem::resize_file(abspath, size);
+  }
+
+  void mkdir() {
+    REQUIRE(vfs_test.vfs_.create_dir(tiledb::sm::URI(abspath.string())).ok());
+  }
+
+  /**
+   * @return a string containing the way this is expected
+   * to appear in the ls_recursive output
+   */
+  std::string lsresult() const {
+    return tiledb::sm::URI(abspath.string()).to_string();
+  }
+
+  bool matches(const std::pair<std::string, uint64_t>& lsout) const {
+    return (lsresult() == lsout.first && size == lsout.second);
+  }
+};
+
+tiledb::sm::LsObjects sort_by_name(const tiledb::sm::LsObjects& in_objs) {
+  tiledb::sm::LsObjects out_objs(in_objs);
+  std::sort(
+      out_objs.begin(), out_objs.end(), [](const auto& f1, const auto& f2) {
+        return f1.first < f2.first;
+      });
+  return out_objs;
+}
+
+TEST_CASE("VFS: ls_recursive unfiltered", "[vfs][ls_recursive]") {
+  std::string prefix = GENERATE("file://");
+  prefix += std::filesystem::current_path().string() + "/ls_recursive_test/";
+
+  VFSTest vfs_test({0}, prefix);
+  const auto mkst = vfs_test.mkdir();
+  REQUIRE(mkst.ok());
+
+  std::vector<TestPath> testpaths = {
+      TestPath(vfs_test, "a1.txt", 30),
+      TestPath(vfs_test, "a2.txt", 40),
+      TestPath(vfs_test, "f1.txt", 10),
+      TestPath(vfs_test, "f2.txt", 20),
+      TestPath(vfs_test, "d1/f1.txt", 45),
+      TestPath(vfs_test, "d1/c1.txt", 55),
+      TestPath(vfs_test, "d1/d1sub1/d1sub1sub1/g1.txt", 33),
+      TestPath(vfs_test, "d1/d1sub1/d1sub1sub1/d1sub1sub1sub1/b1.txt", 12),
+      TestPath(vfs_test, "d1/d1sub1/d1sub1sub1/d1sub1sub1sub1/h1.txt", 33),
+  };
+
+  SECTION("Empty directory") {
+    const auto ls = sort_by_name(vfs_test.vfs_.ls_recursive(
+        vfs_test.temp_dir_,
+        tiledb::sm::accept_all_files,
+        tiledb::sm::accept_all_dirs));
+    CHECK(ls.empty());
+  }
+
+  SECTION("Files only") {
+    testpaths[0].touch();
+    testpaths[1].touch();
+    testpaths[2].touch();
+    testpaths[3].touch();
+
+    const auto ls = sort_by_name(vfs_test.vfs_.ls_recursive(
+        vfs_test.temp_dir_,
+        tiledb::sm::accept_all_files,
+        tiledb::sm::accept_all_dirs));
+    REQUIRE(ls.size() == 4);
+    CHECK(testpaths[0].matches(ls[0]));
+    CHECK(testpaths[1].matches(ls[1]));
+    CHECK(testpaths[2].matches(ls[2]));
+    CHECK(testpaths[3].matches(ls[3]));
+  }
+
+  SECTION("Empty subdirectory") {
+    auto d1 = TestPath(vfs_test, "d1");
+    d1.mkdir();
+
+    const auto ls = sort_by_name(vfs_test.vfs_.ls_recursive(
+        vfs_test.temp_dir_,
+        tiledb::sm::accept_all_files,
+        tiledb::sm::accept_all_dirs));
+
+    CHECK(ls.size() == 1);
```

Comment:
I chose to do it this way because if the `REQUIRE` does not pass, then there is no output whatsoever about what the contents of `ls` actually are.  This way the code is a little uglier but when it fails the first time there is still useful output.  I would prefer not to change this.